source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -48,66 +50,121 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_of
-      ENVOY_LOG(debug, "applying request_headers response");
-      message_valid = true;
-      const auto& headers_response = response->request_headers();
-      if (headers_response.has_response()) {
-        const auto& common_response = headers_response.response();
-        if (common_response.has_header_mutation()) {
-          MutationUtils::applyHeaderMutations(common_response.header_mutation(), *request_headers_);
-        }
-      }
-    } else if (response->has_immediate_response()) {
-      ENVOY_LOG(debug, "Returning immediate response from processor");
-      sendImmediateResponse(response->immediate_response());
-      message_valid = true;
-    }
-    request_state_ = FilterState::IDLE;
-    decoder_callbacks_->continueDecoding();
+  if (response->has_request_headers()) {
+    message_handled = handleRequestHeadersResponse(response->request_headers());
+  } else if (response->has_response_headers()) {
+    message_handled = handleResponseHeadersResponse(response->response_headers());
+  } else if (response->has_immediate_response()) {
+    message_handled = handleImmediateResponse(response->immediate_response());
   }
 
-  if (message_valid) {
+  if (message_handled) {
     stats_.stream_msgs_received_.inc();
   } else {
     stats_.spurious_msgs_received_.inc();
     // Ignore messages received out of order. However, close the stream to
     // protect ourselves since the server is not following the protocol.
     ENVOY_LOG(warn, "Spurious response message received on gRPC stream");
+    cleanupState();
     closeStream();
   }
 }
 
+bool Filter::handleRequestHeadersResponse(const HeadersResponse& response) {
+  if (request_state_ == FilterState::HEADERS) {
+    ENVOY_LOG(debug, "applying request_headers response");
+    if (response.has_response()) {
+      const auto& common_response = response.response();
+      if (common_response.has_header_mutation()) {
+        MutationUtils::applyHeaderMutations(common_response.header_mutation(), *request_headers_);
+      }
+    }
+    request_state_ = FilterState::IDLE;
+    decoder_callbacks_->continueDecoding();
+    return true;
+  }
+  return false;
+}
+
+bool Filter::handleResponseHeadersResponse(const HeadersResponse& response) {
+  if (response_state_ == FilterState::HEADERS) {
+    ENVOY_LOG(debug, "applying response_headers response");
+    if (response.has_response()) {
+      const auto& common_response = response.response();
+      if (common_response.has_header_mutation()) {
+        MutationUtils::applyHeaderMutations(common_response.header_mutation(), *response_headers_);
+      }
+    }
+    response_state_ = FilterState::IDLE;
+    encoder_callbacks_->continueEncoding();
+    return true;
+  }
+  return false;
+}
+
+bool Filter::handleImmediateResponse(const ImmediateResponse& response) {
+  if (response_state_ == FilterState::HEADERS) {
+    // Waiting for a response headers response, so return immediately now.
+    // Do this first in case both are in progress.
```

Comment:
As an observation, the decoupling of request/response state from server reply message type makes it a bit confusing to reason about the state machine here. If we had distinct message for "send immediate response" for request/reply, it would be a bit more intuitive in the code at least; OTOH, I think the behaviors are identical so it's technically not needed.

Reply:
I'm not entirely sure how to improve this area. For this PR I switched things up to have one function that handles the incoming message and switches on the message type, and others that decide whether it's an appropriate time to handle each one. Or are you suggesting changing the proto so that "immediate_response" is part of a larger message type rather than a separate message?

Reply:
I guess tracking the containing context (request/response) for the immediate response would be clearer than relying on request/response state variables to me, but I can see either working.